### PR TITLE
Evaluate Dart expressions by sending them to the server for compilation

### PR DESCRIPTION
**Pull request by alan-knight**
I apologize for the size of this. It also still has numerous holes.
  - part files don't work at all
  - evaluating inside closures in methods/functions doesn't work (it assumes a single level)
  - the initial evaluation is very slow (has to start ddc on the server)
  - variables imported in other libraries aren't visible
  - only single expressions work
  - it only works in the console so far
But it's still a very large improvement.

### CODE REVIEW

File: front_end/dart_tests/context_for_evaluation.dart [hunk 1]
```diff
@@ -0,0 +1,24 @@
+// Context for tests to exercise evaluating expressions.
+
```
Comment (jacob314) [inline front_end/dart_tests/context_for_evaluation.dart:2]: add a TODO calling out additional dart specific edge cases that should be tested here.
For example:
imports referenced with a prefix, names for cases where DDC has to perform renaming.

File: front_end/dart_tests/dart_support_test.dart [hunk 2]
```diff
@@ -1,45 +1,170 @@
...
-
-@JS()
 library dart_support_test;
 
 @TestOn('chrome')
 import 'package:test/test.dart';
 import 'package:js/js.dart';
 import 'dart:html';
 import 'dart:async';
+// TODO: Would it be better to pre-compile this and load it from a script tag?
```
Comment (jacob314) [inline front_end/dart_tests/dart_support_test.dart:8]: TODO missing a username

File: front_end/dart_tests/dart_support_test.dart [hunk 3]
```diff
@@ -1,45 +1,170 @@
...
+  test("Not Dart code bails out", () {
+    unmodified('(a/&*');
   });
 
-  test("Doesn't look like Dart", () {
-    unmodified('a/&*');
+  test("Basic evaluating in a context", () {
+    // The 'context' library has hard-coded evaluation points. Make sure that we
+    // can evaluate a simple expression.
+    jsSetThingToEval('2+2');
```
Comment (jacob314) [inline front_end/dart_tests/dart_support_test.dart:57]: nit: make your basic expression something that wouldn't work with JS eval.
e.g. test that
'a' * 3
returns 'aaa'

File: front_end/bindings/BlackboxManager.js [hunk 4]
```diff
@@ -341,7 +341,63 @@ Bindings.BlackboxManager = class {
     if (!script.sourceURL && !script.sourceMapURL)
       return Promise.resolve();
     const blackboxed = this._isBlackboxedScript(script);
-    return this._setScriptState(script, blackboxed ? [{lineNumber: 0, columnNumber: 0}] : []);
+    var ranges =  blackboxed ? [{lineNumber: 0, columnNumber: 0}] : [];
+    if (script.sourceURL.includes('dart_sdk.js')) {
```
Comment (jacob314) [inline front_end/bindings/BlackboxManager.js:345]: nit: can you be a bit more strict and make this endsWith '/dart_sdk.js?'

File: front_end/dart_tests/dart_support_test.dart [hunk 5]
```diff
@@ -1,45 +1,170 @@
...
-main() async {
+/// Create a script element for [path] to load it.
+ScriptElement loadScript(String path) {
+  // Note that we have to do this in a separate non-async method
+  // because of https://github.com/dart-lang/sdk/issues/35020
+  var element = new ScriptElement()
+    ..type = 'text/javascript'
+    ..async = false
+    ..src = path;
+  document.body.append(element);
```
Comment (jacob314) [inline front_end/dart_tests/dart_support_test.dart:27]: nit: isn't appending to the header instead of the body better style?

File: front_end/dart_tests/test_helpers.dart [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+class FakeExecutionContext {
+  @JS('selectedCallFrame')
+  external set selectedCallFrame(x);
+}
+
+/// A fake that we use to replace the devtools Bindings namespace.
+class FakeChromeBindings {
+  get debuggerWorkspaceBinding => this;
+  UILocation _location = UILocation();
+  rawLocationToUILocation(_) => _location;
```
Comment (jacob314) [inline front_end/dart_tests/test_helpers.dart:155]: why isn't this a getter called location?

File: front_end/dart_tests/test_helpers.dart [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+  get sourceURL => 'someURL';
+
+  /// If true we retain scopeChainForWeb, otherwise scopeChainForLib.
+  bool inWeb;
+  scopeChain() => inWeb ? scopeChainWeb() : scopeChainLib();
+
+  /// Allow creating a call frame in one of two flavors, one hard-coded to a
+  /// particular file in the web directory, and another in lib.
+  FakeCallFrame(this.inWeb);
+  FakeCallFrame.inWeb() : this.inWeb = true;
```
Comment (jacob314) [inline front_end/dart_tests/test_helpers.dart:190]: remove extra this. You have been writing too much JS :)

File: front_end/dart_tests/test_helpers.dart [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+  FakeCallFrame.inLib() : this.inWeb = false;
+
+  /// We don't actually evaluate, the test sets the desired result.
+  var evaluationResult;
+  Promise evaluate(evaluateOptions, userGesture, awaitPromise) {
+    return futureToPromise(Future.value(evaluationResult));
+  }
+
+  /// A hard-coded scope chain for a sample file in the web directory.
+  scopeChainWeb() => [
```
Comment (jacob314) [inline front_end/dart_tests/test_helpers.dart:200]: nit: can this just be
final scopeChainWeb = [....]

File: front_end/dart_tests/test_helpers.dart [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+  get _webFunctionName => 'foo__bar__web__main';
+  get _libFunctionName => 'foo__bar__library';
+  get _functionName => inWeb ? _webFunctionName : _libFunctionName;
+  get functionName => 'dart_library.library.$_functionName.actualName';
+}
+
+/// Fake for a devtools "Bindings". This is not the same as the Bindings
+/// namespace, it's an anonymous object with a name and properties, where the
+/// properties are SDK.RemoteObjectProperties.
+class Bindings {
```
Comment (jacob314) [inline front_end/dart_tests/test_helpers.dart:236]: can you provide a few more types for clarity for this code. E.g. what is the type of _name?

File: front_end/sdk/DartEval.js [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+    /// The name of the DDC module that the current callFrame is in.
+    ///
+    /// We do this by finding the URL of the current frame and then looking into
+    /// the $dartLoader data structures in the application.  TODO: Make this
+    /// less dependent on the module system, or have a version for AMD.
+    async _currentModuleId() {
+        const url = this.callFrame.functionLocation().script().sourceURL;
+        const module = await this.callFrame.evaluate(
+            {
+                expression: '$dartLoader.urlToModuleId.get("' + url + '")',
```
Comment (jacob314) [inline front_end/sdk/DartEval.js:210]: url also needs to be escaped.

File: front_end/sdk/DartEval.js [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+                silent: true,
+                returnByValue: true,
+                generatePreview: false
+            },
+              /* userGesture */ false,
```
Comment (jacob314) [inline front_end/sdk/DartEval.js:215]: indent seems off

File: front_end/sdk/DartEval.js [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+        const environments = await this.environments();
+        const allNames = [];
+        for (const scope of environments) {
+            var prefixed;
+            if (scope.prefix) {
+                prefixed = scope.items.map(thing => scope.prefix + '.' + thing);
+            } else {
+                prefixed = scope.items;
+            }
+            allNames.push(...prefixed); // TODO: shadowing
```
Comment (jacob314) [inline front_end/sdk/DartEval.js:234]: clarify this comment and add your username

File: front_end/sdk/DartSupport.js [hunk 13]
```diff
@@ -1,6 +1,8 @@
+// Functions specific to Dart support.
```
Comment (jacob314) [inline front_end/sdk/DartSupport.js:1]: nit: clarify the comment. Seems like this a module specific to Dart support with everything added under the Dart namespace.

File: front_end/sdk/DartSupport.js [hunk 14]
```diff
@@ -56,78 +58,72 @@ Dart.ToggleSourcemapsActionDelegate = class {
...
+/// Convert a Dart expression to JavaScript.
+///
+/// We wrap the expression in a function which provides the local scope
+/// variables. That lets us evaluate the expression as a standalone function and
+/// not try to evaluate in the right scope.
+///
+/// @param {SDK.ExecutionContext} executionContext
+/// @param {string} dartExpression
+/// @return {string} dartExpression compiled to JavaScript
+window.$dartExpressionFor = async function (executionContext, dartExpression) {
```
Comment (jacob314) [inline front_end/sdk/DartSupport.js:70]: why is this logic written in JavaScript? Seems better to write in Dart now that more logic is in dart.

**Review by jacob314 — APPROVED** [review]
lgtm with nits

File: front_end/dart_tests/context_for_evaluation.dart (hunk 1 above)
Comment (alan-knight) [reply front_end/dart_tests/context_for_evaluation.dart:2]: Done, in the main test file.

File: front_end/dart_tests/dart_support_test.dart (hunk 3 above)
Comment (alan-knight) [reply front_end/dart_tests/dart_support_test.dart:57]: At this point it's actually evaluating JavaScript. Added a comment and changed the parameter name to clarify this.

File: front_end/bindings/BlackboxManager.js (hunk 4 above)
Comment (alan-knight) [reply front_end/bindings/BlackboxManager.js:345]: That requires parsing the URL so we can get just the path part. Did it by doing this check first, and then the more strict check inside, since this gets called for every file.

File: front_end/dart_tests/dart_support_test.dart (hunk 2 above)
Comment (alan-knight) [reply front_end/dart_tests/dart_support_test.dart:8]: Done throughout.

File: front_end/dart_tests/dart_support_test.dart (hunk 5 above)
Comment (alan-knight) [reply front_end/dart_tests/dart_support_test.dart:27]: Done.

File: front_end/dart_tests/test_helpers.dart (hunk 7 above)
Comment (alan-knight) [reply front_end/dart_tests/test_helpers.dart:190]: Done.

File: front_end/dart_tests/test_helpers.dart (hunk 6 above)
Comment (alan-knight) [reply front_end/dart_tests/test_helpers.dart:155]: It's trying to mirror what the devtools Bindings does, and gets called from JS.

Comment (jacob314) [reply front_end/dart_tests/test_helpers.dart:155]: sg

File: front_end/dart_tests/test_helpers.dart (hunk 8 above)
Comment (alan-knight) [reply front_end/dart_tests/test_helpers.dart:200]: Yes, that's good. Those used to be destructively modified, but they're not any more, so they can be. Done.

File: front_end/dart_tests/test_helpers.dart (hunk 9 above)
Comment (alan-knight) [reply front_end/dart_tests/test_helpers.dart:236]: Done.

File: front_end/sdk/DartEval.js (hunk 12 above)
Comment (alan-knight) [reply front_end/sdk/DartEval.js:234]: Done.

File: front_end/sdk/DartEval.js (hunk 11 above)
Comment (alan-knight) [reply front_end/sdk/DartEval.js:215]: Done.

File: front_end/sdk/DartSupport.js (hunk 13 above)
Comment (alan-knight) [reply front_end/sdk/DartSupport.js:1]: Done.

File: front_end/sdk/DartSupport.js (hunk 14 above)
Comment (alan-knight) [reply front_end/sdk/DartSupport.js:70]: That would be nice, especially since it includes logic similar to what might be needed to implement VM protocol. But it's tricky, as this is called from devtools and runs inside it, doing a bunch of back and forth on the protocol. So it's hard to move it to the server,  and right now we don't have any Dart code included in DDT. I'm not sure where the ability to run Dart code living inside a JS main is, but it at least would be tricky. Added a TODO for it.

Comment (jacob314) [reply front_end/sdk/DartSupport.js:70]: sg

File: front_end/sdk/DartEval.js (hunk 10 above)
Comment (alan-knight) [reply front_end/sdk/DartEval.js:210]: JSON escaped it.
